typing/typecore.ml

```diff
@@ -1906,18 +1901,19 @@ and type_pat_aux
         }
       in
       let k' pat = rvp k @@ solve_expected pat in
-      begin match mode with
-      | Normal ->
-          k' (wrap_disambiguate "This record pattern is expected to have"
-               (mk_expected expected_ty)
-               (type_label_a_list loc false !env Env.Projection
-                  type_label_pat expected_type lid_sp_list)
-               make_record_pat)
-      | Counter_example {labels; _} ->
-          type_label_a_list ~labels loc false !env Env.Projection
-            type_label_pat expected_type lid_sp_list
-            (fun lbl_pat_list -> k' (make_record_pat lbl_pat_list))
-      end
+      let lbl_a_list =
+        match mode with
+        | Normal ->
+            wrap_disambiguate
+              "This record pattern is expected to have"
+              (mk_expected expected_ty)
+              (disambiguate_record loc false !env Env.Projection expected_type)
+              lid_sp_list
+        | Counter_example {labels; _} ->
+            disambiguate_record ~labels loc false !env Env.Projection
+              expected_type lid_sp_list
+      in
```

Comment:
Question reading the code (not related to th change directly): why are we using `wrap_disambiguate` only in the `Normal` case? I think this corresponds to an assumption about the shape of `Counter_example` labels (but if this cannot fail, why are we calling `disambiguate_record` in this branch?). It may be clearer to wrap the whole definition of `lbl_a_list`, to ensure that all calls to `disambiguate_*` are wrapped?

Reply:
The role of `wrap_disambiguate` is to provide an error message.
Since we do not report errors in `Counter_example` mode (just backtrack), there is no need to use it.
This said, I would expect to be no need to call `disambiguate_record` in this mode, since the labels are already given. Probably there is code in `disambiguate_record` that should rather be here.